logot/_structlog.py

```diff
@@ -0,0 +1,76 @@
+from __future__ import annotations
+
+from functools import partial
+from typing import Any, MutableMapping
+
+import structlog
+from structlog.exceptions import DropEvent
+
+from logot._capture import Captured
+from logot._logot import Capturer, Logot
+from logot._typing import Level, Name
+
+EventDict = MutableMapping[str, Any]
+WrappedLogger = Any
+
+
+CRITICAL = 50
+FATAL = CRITICAL
+ERROR = 40
+WARNING = 30
+WARN = WARNING
+INFO = 20
+DEBUG = 10
+NOTSET = 0
+
+NAME_TO_LEVEL = {
+    "critical": CRITICAL,
+    "exception": ERROR,
+    "error": ERROR,
+    "warn": WARNING,
+    "warning": WARNING,
+    "info": INFO,
+    "debug": DEBUG,
+    "notset": NOTSET,
+}
+
+
+class StructlogCapturer(Capturer):
+    """
+    A :class:`logot.Capturer` implementation for :mod:`structlog`.
+    """
+
+    __slots__ = ("_old_processors", "_old_wrapper_class")
+
+    def start_capturing(self, logot: Logot, /, *, level: Level, name: Name) -> None:
+        config = structlog.get_config()
+        processors = config["processors"]
+        wrapper_class = config["wrapper_class"]
+        self._old_processors = processors.copy()
+        self._old_wrapper_class = wrapper_class
+        processors.clear()
+        processors.append(partial(_processor, logot=logot, name=name))
```

Comment:
I'm a little wary of of mutating the returned processors list. What if instead, we stored the returned processor list unchanged, and configured `struct`log with a new one? This applies to restoring the original state... we can just put the original processor list back unmutated.

We should also be preserving the existing processors. That allows multiple stacked calls to `capture()` to coexist.

So something like 

```py
self._old_processors = processors
```

And 
```py
structlog.configure(
    processors=[*processors, partial(_processor, logot=logot, name=name)],
    wrapper_class=wrapper_class,
)
```

Will that work?

Reply:
I did not spend enough time reading the existing implementations / docs here! This does work. However it depends on some nuances. Thanks for clarifying above

> A capturer is supposed to leave any other log processing as-is, rather than intercept and squash. This allows multiple capture calls to coexist.

I'd like to be clear whether the expectation for logot is to transparently intercept logs at

1. The call site
2. The log "sink" (eg. after all processing and filtering is done)

As with structlog, this has a unique distinction. Processors can (as you can see here) filter events. If we insert the `_processor` at the end of the chain, we may miss some logs that have been emitted due to arbitrary logic in the chain. However, if we insert it at the beginning, we will get logs without additional or transformed fields - their "raw state". This may affect https://github.com/etianen/logot/issues/25

Either way I feel this is something that should be made clear in the documentation (probably in the "log capturing" section) - let me know your thoughts!

Reply:
That's interesting. I'm inclined to say we should go at the end of the chain then. If people are filtering logs in their processors, then that also lets us test that filtering.

I'd agree that this should be mentioned in the docs!

There's also an interesting thing here that `structlog` might benefit from some additional optional configuration being passed to the `Capturer`. I did experiment with this before using `ParamSpec`, but it got quite nasty with typing. I might revisit this again if it turns out putting the log capture at the end of the chain proves a problem.

Reply:
Note from the changes; the processor has to be inserted second-to-last, to avoid interfering with the job of the final processor https://www.structlog.org/en/stable/processors.html#adapting-and-rendering

I've called this out in the documentation